test/tbb/test_join_node_preview.cpp

```diff
@@ -0,0 +1,98 @@
+
+    std::array<msg_t, 3> messages_for_follows = { {msg_t(), msg_t(), msg_t()} };
+    std::vector<msg_t> messages_for_precedes = {msg_t(), msg_t(), msg_t()};
+
+    follows_and_precedes_testing::test_follows
+        <msg_t, tbb::flow::join_node<JoinOutputType>, tbb::flow::buffer_node<msg_t>>(messages_for_follows);
+    follows_and_precedes_testing::test_follows
+        <msg_t, tbb::flow::join_node<JoinOutputType, tbb::flow::queueing>>(messages_for_follows);
+    follows_and_precedes_testing::test_follows
+        <msg_t, tbb::flow::join_node<JoinOutputType, tbb::flow::reserving>, tbb::flow::buffer_node<msg_t>>(messages_for_follows);
+    auto b = [](msg_t) { return msg_t(); };
+    class hash_compare {
+    public:
+        std::size_t hash(msg_t) const { return 0; }
+        bool equal(msg_t, msg_t) const { return true; }
+    };
+    follows_and_precedes_testing::test_follows
+        <msg_t, tbb::flow::join_node<JoinOutputType, tbb::flow::key_matching<msg_t, hash_compare>>, tbb::flow::buffer_node<msg_t>>
+        (messages_for_follows, b, b, b);
+
+    follows_and_precedes_testing::test_precedes
+        <msg_t, tbb::flow::join_node<JoinOutputType>>(messages_for_precedes);
+    follows_and_precedes_testing::test_precedes
+        <msg_t, tbb::flow::join_node<JoinOutputType, tbb::flow::queueing>>(messages_for_precedes);
+    follows_and_precedes_testing::test_precedes
+        <msg_t, tbb::flow::join_node<JoinOutputType, tbb::flow::reserving>>(messages_for_precedes);
+    follows_and_precedes_testing::test_precedes
+        <msg_t, tbb::flow::join_node<JoinOutputType, tbb::flow::key_matching<msg_t, hash_compare>>>
+        (messages_for_precedes, b, b, b);
+}
+
+void jn_msg_key_matching_follows_and_precedes() {
+    using msg_t = MyMessageKeyWithoutKey<int, int>;
+    using JoinOutputType = std::tuple<msg_t, msg_t, msg_t>;
+
+    std::array<msg_t, 3> messages_for_follows = { {msg_t(), msg_t(), msg_t()} };
+    std::vector<msg_t> messages_for_precedes = { msg_t(), msg_t(), msg_t() };
+
+    follows_and_precedes_testing::test_follows
+        <msg_t, tbb::flow::join_node<JoinOutputType, tbb::flow::key_matching<std::size_t>>, tbb::flow::buffer_node<msg_t>>
+        (messages_for_follows);
+    follows_and_precedes_testing::test_precedes
+        <msg_t, tbb::flow::join_node<JoinOutputType, tbb::flow::key_matching<std::size_t>>>
+        (messages_for_precedes);
+}
+
+void test_follows_and_precedes_api() {
+    jn_follows_and_precedes();
+    jn_msg_key_matching_follows_and_precedes();
+}
+
+//! Test follows and precedes API
+//! \brief \ref error_guessing
+TEST_CASE("Test follows and preceedes API"){
+    test_follows_and_precedes_api();
+}
+
+//! Test maintaining correct count of ports without input
+//! \brief \ref error_guessing
+TEST_CASE("Test removal of the predecessor while having none") {
+    using namespace multiple_predecessors;
+
+    test(connect_join_via_follows);
+    test(connect_join_via_precedes);
```

Comment:
Is it (`follows` and `precedes`)-specific?

Reply:
Do you suggest move it to the test case "Test follows and precedes API"?

Reply:
No, I am trying to understand if this test is preview-specific or not, and judging by the name of the test case, it should not be. However, we need to check that. And, if it is not preview-specific, I suggest rewriting it by not using preview features, and then move out of this file to better separate preview and non-preview testing of the join node.

Reply:
I believe it is because functions connect_join_via_follows and connect_join_via_precedes contain calls of `follows(...)` and `precedes(...)` functions. And in original testing this part was also under PREVIEW macro.

Reply:
Decided to leave a TODO about further analysis into this.